### ISSUE TITLE
Tried to write class-balanced cross entropy, but couldn't understand it. 

### ISSUE THREAD

**Issue by GodOfSmallThings** [argman/EAST#12]
I just want to modify loss function, from dice coefficient to class balanced xentropy, but I still don't get what to change.

def batch_flatten(x):
    """
    Flatten the tensor except the first dimension.
    """
    shape = x.get_shape().as_list()[1:]
    if None not in shape:
        return tf.reshape(x, [-1, int(np.prod(shape))])
    return tf.reshape(x, tf.stack([tf.shape(x)[0], -1]))

def xentropy(y_true_cls, y_pred_cls,
                     training_mask):
    eps = 1e-7
    
    z = batch_flatten(y_pred_cls)
    y = tf.cast(batch_flatten(y_true_cls), tf.float32)

    count_neg = tf.reduce_sum((1. - y) * training_mask)
    count_pos = tf.reduce_sum(y * training_mask)
    beta = count_neg / (count_neg + count_pos)
    loss_pos = -beta * tf.reduce_mean(y * tf.log(z + eps))
   loss_neg = (1. - beta) * tf.reduce_mean((1. - y) * tf.log(1. - z + eps))
   cost = tf.subtract(loss_pos, loss_neg, name=name)

 return cost
---------------------------------------------------------------------

Would this code be work?

**Comment by argman** [comment]
I donot recommend implement xentropylike this, you can tf.nn.weighted_corss_entropy_with_logits

**Comment by GodOfSmallThings** [comment]
I need to implement class-balanced cross entropy function. That's why I'm digging into it !

**Comment by GodOfSmallThings** [comment]
So if I want to implement class-balanced xentropy, is it okay to ignore training mask part?

**Comment by argman** [comment]
You can use this one.
```
def weighted_cross_entropy(y_true_cls, y_pred_cls,
                           training_mask):
    # Loss for score Map
    epsilon = 10e-8
    # assume ratio of positive class is 0.1
    beta = 1.0 - 0.1
    # turn probabilities back to logits
    y_pred_cls = tf.clip_by_value(y_pred_cls, epsilon, 1 - epsilon)
    y_pred_cls = tf.log(y_pred_cls / (1. - y_pred_cls))
    Ls = (1. - beta) * tf.nn.weighted_cross_entropy_with_logits(targets=y_true_cls, logits=y_pred_cls,
                                                                pos_weight=beta / (1. - beta))
    Ls = tf.reduce_mean(Ls * training_mask)
    tf.summary.scalar('classification_logloss', Ls)
    return Ls
```

**Comment by GodOfSmallThings** [comment]
I 'll try. Thank you very much for the answer!
